### PR TITLE
WIP: RxFire

### PR DESCRIPTION
**Pull request by davideast**
### Discussion

RxJS + Firebase = 🔥🔥🔥

WIP. Will provide a deep description soon.

### CODE REVIEW

File: packages/rxfire/firestore/collection/index.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+      }
+      break;
+    case 'modified':
+      // When an item changes position we first remove it
+      // and then add it's new position
+      if (change.oldIndex !== change.newIndex) {
+        combined.splice(change.oldIndex, 1);
+        combined.splice(change.newIndex, 0, change);
+      } else {
+        combined.splice(change.newIndex, 1, change);
```
Comment (jshcrowthe) [inline packages/rxfire/firestore/collection/index.ts]: This is just a more difficult way of saying:

```typescript
combined[change.newIndex] = change;
``` 

Can we swap it out and add a comment?

File: packages/rxfire/firestore/collection/index.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  (changes: firestore.DocumentChange[]) => changes.length > 0
+);
+
+/**
+ * Creates a new sorted array from a new change.
+ * @param combined
+ * @param change
+ */
+function combineChange<T>(
+  combined: firestore.DocumentChange[],
```
Comment (jshcrowthe) [inline packages/rxfire/firestore/collection/index.ts]: Let's be more descriptive of what this actually is in this name (i.e. the full array of changes and a single change to process)

File: packages/rxfire/firestore/collection/index.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+const filterEmpty = filter(
+  (changes: firestore.DocumentChange[]) => changes.length > 0
+);
+
+/**
+ * Creates a new sorted array from a new change.
+ * @param combined
+ * @param change
+ */
+function combineChange<T>(
```
Comment (jshcrowthe) [inline packages/rxfire/firestore/collection/index.ts]: This function name isn't indicative of what is actually being done here, let's think of a new name?

File: packages/rxfire/firestore/collection/index.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+) {
+  return fromCollectionRef(query).pipe(
+    map(snapshot => snapshot.docChanges()),
+    filterEvents(events),
+    filterEmpty
+  );
+}
+
+/**
+ * Return a stream of document changes on a query. These results are in sort order.
```
Comment (jshcrowthe) [inline packages/rxfire/firestore/collection/index.ts]: This comment is inaccurate, we return a stream of `DocumentSnapshots` not changes.

File: packages/rxfire/firestore/collection/index.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        changes: firestore.DocumentChange[]
+      ) => combineChanges(current, changes, events),
+      []
+    )
+  );
+}
+
+/**
+ * Create a stream of changes as they occur it time. This method is similar
+ * to stateChanges() but it collects each event in an array over time.
```
Comment (jshcrowthe) [inline packages/rxfire/firestore/collection/index.ts]: `stateChanges` doesn't exist?

File: packages/rxfire/storage/index.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  data: any,
+  metadata?: storage.UploadMetadata
+) {
+  return fromTask(ref.put(data, metadata));
+}
+
+export function putString(
+  ref: storage.Reference,
+  data: string,
+  format?: storage.StringFormat,
```
Comment (jshcrowthe) [inline packages/rxfire/storage/index.ts]: This isn't used in the body? Perhaps you meant `ref.putString`?

File: packages/rxfire/firestore/collection/index.ts (hunk 5 above)
Comment (davideast) [reply packages/rxfire/firestore/collection/index.ts]: it's now `docChanges()`

File: packages/rxfire/storage/index.ts (hunk 6 above)
Comment (davideast) [reply packages/rxfire/storage/index.ts]: fixed

File: packages/rxfire/firestore/collection/index.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+/**
+ * Create an operator that determines if a the stream of document changes
+ * are specified by the event filter. If the document change type is not
+ * in specified events array, it will not be emitted.
+ */
+const filterEvents = (events?: firestore.DocumentChangeType[]) =>
+  filter((changes: firestore.DocumentChange[]) => {
+    let hasChange = false;
+    changes.forEach(change => {
```
[the remark this answers is not in the document]
Comment (merlinnot) [reply packages/rxfire/firestore/collection/index.ts]: Or even [some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some)?

File: packages/rxfire/firestore/collection/index.ts (hunk 4 above)
Comment (davideast) [reply packages/rxfire/firestore/collection/index.ts]: fixed.

File: packages/rxfire/firestore/collection/index.ts (hunk 3 above)
Comment (davideast) [reply packages/rxfire/firestore/collection/index.ts]: fixed.

File: packages/rxfire/firestore/collection/index.ts (hunk 2 above)
Comment (davideast) [reply packages/rxfire/firestore/collection/index.ts]: fixed.

File: packages/rxfire/firestore/collection/index.ts (hunk 1 above)
Comment (davideast) [reply packages/rxfire/firestore/collection/index.ts]: fixed

File: packages/rxfire/firestore/collection/index.ts [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  combined: firestore.DocumentChange[],
+  change: firestore.DocumentChange
+): firestore.DocumentChange[] {
+  switch (change.type) {
+    case 'added':
+      if (
+        combined[change.newIndex] &&
+        combined[change.newIndex].doc.id == change.doc.id
+      ) {
+        // Not sure why the duplicates are getting fired
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/firestore/collection/index.ts]: fixed

File: packages/rxfire/firestore/document/index.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { firestore } from 'firebase/app';
+import { Observable } from 'rxjs';
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/firestore/document/index.ts]: fixed

File: packages/rxfire/firestore/fromRef.ts [hunk 10]
```diff
@@ -0,0 +1,40 @@
+/**
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/firestore/fromRef.ts]: action item for SDK

File: packages/rxfire/firestore/fromRef.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { firestore } from 'firebase/app';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/firestore/fromRef.ts]: fixed

File: packages/rxfire/firestore/fromRef.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+type Reference<T> = firestore.DocumentReference | firestore.CollectionReference;
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/firestore/fromRef.ts]: fixed

File: packages/rxfire/firestore/fromRef.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+function _fromRef<T, R>(ref: any): Observable<R> {
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/firestore/fromRef.ts]: fixed

File: packages/rxfire/rollup.config.js [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+import commonjs from 'rollup-plugin-commonjs';
+import typescript from 'rollup-plugin-typescript2';
+import uglify from 'rollup-plugin-uglify';
+import pkg from './package.json';
+
+import authPkg from './auth/package.json';
+import storagePkg from './storage/package.json';
+import functionsPkg from './functions/package.json';
+
+const pkgsByName = {
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/rollup.config.js]: fixed

File: packages/rxfire/rollup.config.js [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+];
+
+const external = [...Object.keys(pkg.dependencies || {}), 'rxjs/operators'];
+
+/**
+ * Global UMD Build
+ */
+const GLOBAL_NAME = 'rxfire';
+
+const coreBuild = [
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/rollup.config.js]: fixed

File: packages/rxfire/rollup.config.js [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        external: ['firebase', 'rxjs', 'rxjs/operators']
+      }
+    ];
+  })
+  .reduce((a, b) => a.concat(b), []);
+
+/**
+ * Complete Package Builds
+ */
+const completeBuilds = [
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/rollup.config.js]: fixed

File: packages/rxfire/package.json [hunk 17]
```diff
@@ -0,0 +1,69 @@
+{
+  "name": "rxfire",
+  "version": "0.0.2",
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/package.json]: fixed

File: packages/rxfire/auth/package.json [hunk 18]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "rxfire-auth",
```
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/auth/package.json]: fixed

File: packages/rxfire/firestore/collection/index.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (davideast) [reply packages/rxfire/firestore/collection/index.ts]: fixed

**Pull request merged**